### ISSUE TITLE
Should we use [EnforceRange] on min/max in constraints?

### ISSUE THREAD

**Issue by alvestrand** [w3c/mediacapture-main#384]
I was trying to debug a test with this constraint:

navigator.getUserMedia({video: {width: {min:Infinity}}} ....)

It failed in an interesting way (so there is probably a bug), but I tried to figure out what SHOULD happen.

WebIDL: https://www.w3.org/TR/WebIDL/#es-long - short version: ToInt32(ToNumber(string))

Assuming that "Infinity" is spelled correctly, this is ToInt32(+inf). Follow link to ECMAScript.

http://es5.github.io/#x9.5 - If number is NaN, +0, −0, +∞, or −∞, return +0. (WAT)

So if it's just a very large number? "Let int32bit be posInt modulo 2^32"

Result: constraining to +inf is the same as constraining to 0. Constraining to 2^32+100 is the same as constraining to 100.

Would the behavior specified at https://www.w3.org/TR/WebIDL/#es-long for [EnforceRange] (which is "throw a TypeError") be a less surprising behavior for users?

**Comment by dontcallmedom** [comment]
I agree that `[EnforceRange]` would work better; alternatively, `[Clamp]`would not throw, but instead clamp down to the closest acceptable value.

(whichever we choose, we would have to apply it to:
* `min`, `max` in `LongRange`
*`ideal`, `exact` in `ConstrainLongRage`
* the `long` type in the `ConstrainLong` typedef (but I'm not sure if it's syntactically valid — the WebIDL specs doesn't talk about applying it to union types)

I noticed that all the current usages of ConstrainLong are for positive values only; I'll raise a separate issue to discuss with we should be using `unsigned long` instead.

**Comment by alvestrand** [comment]
The TAG advice at https://w3ctag.github.io/design-principles/#numeric-types suggests using [EnforceRange] long long.
I don't see a powerful reason to change the interface type, but it's a nudge towards preferring [EnforceRange] over [Clamp].

(Amusement: WebIDL "double" throws TypeError on being passed infinity. WebIDL "long" doesn't.)

**Comment by adam-be** [comment]
I've started a thread on public-script-coord on how to use [EnforceRange] with the union type and enum. Hopefully we can get some clarity soon.

**Comment by adam-be** [comment]
Some progress have been made [1]

[1] https://lists.w3.org/Archives/Public/public-script-coord/2016JulSep/0033.html

**Comment by alvestrand** [comment]
Waiting for WebIDL progress.

**Comment by jan-ivar** [comment]
> Amusement: WebIDL "double" throws TypeError on being passed infinity.

That's because we use `double` and not `unrestricted double`. We could fix that at the same time.

**Comment by jan-ivar** [comment]
Given that `min` and `max` define boundaries, and `ideal` has gravity, it might be worth considering `[Clamp]` for `long` values, since `-Infinity` and `Infinity` would then produce −2147483648 and 2147483647 respectively, which in the constraints algorithm seem indistinguishably large enough for most imaginable measures of hardware.

This would let users use `Infinity` and `-Infinity` in JS as shorthands for "unconstrained":

    let max = Infinity, min = -Infinity;
    for (c of otherConcerns) {
      if (max > c.max) max = c.max;
      if (min < c.min) min = c.min;
    }
    getUserMedia({video: {min, max}).then(useStream)

Also, `{ width: Infinity, height: Infinity}` would give users the highest available resolution for quite a while.

`unrestricted double` would do the same for `frameRate`.

**Comment by jan-ivar** [comment]
See also https://lists.w3.org/Archives/Public/public-media-capture/2014Jul/0093.html

**Comment by dontcallmedom** [comment]
With [extended attributes now applicable to types](https://github.com/heycam/webidl/issues/286) in WebIDL level 2, it seems the following would solve this issue and #385:
* make `ContrainLongRange.min`, `ContrainLongRange.max`, `ContrainLong.ideal`, `ContrainLongRange.exact` of type `[Clamp] unsigned long`
* make the typedef of `ConstrainLong` to be `([Clamp] unsigned long or ConstrainLongRange)`

before a mergeable PR can be made to that effect, [webidl2.js](https://github.com/w3c/webidl2.js/) and ReSpec needs to be updated to support the new construction. (I have already [updated widlproc](https://github.com/dontcallmedom/widlproc/issues/18) which we also use in our CI checks).

The reasons I'm suggesting this compared to what was discussed above:
* I think `[Clamp]` is more friendly since as @jan-ivar points out, it allows to use `Infinity` as a fallback value for `max`
* I think we should use `unsigned` (and thus disallow `-Infinity` for `min`) since the values can only be positive (and thus the default fallback for `min` can be 0)

**Comment by alvestrand** [comment]
Something came up that caused me to re-look at this issue. @dontcallmedom is webIDL support now in a place where we can take this one out of the icebox?

**Comment by dontcallmedom** [comment]
yes, I believe a pull request should be doable now; I'll take a stab at it

**Comment by dontcallmedom** [comment]
I've prepared a pull request, but ReSpec will still need an upgrade before we can merge:
 https://github.com/w3c/respec/issues/1574
